File: packages/control-plane/loop_control_plane/incidents.py
from __future__ import annotations

import asyncio
from datetime import UTC, datetime
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field

from loop_control_plane._app_agents import AgentRecord
from loop_control_plane.workspaces import WorkspaceError

IncidentSeverity = Literal["low", "medium", "high", "critical"]
IncidentStatus = Literal[
    "open",
    "contained",
    "investigating",
    "fix_staged",
    "resolved",
    "archived",
]


class IncidentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    deployment_id: str = Field(default="", max_length=160)
    severity: IncidentSeverity = "high"
    trigger: str = Field(min_length=1, max_length=500)
    affected_trace_ids: list[str] = Field(default_factory=list, max_length=100)
    affected_conversation_count: int = Field(default=0, ge=0)
    root_cause_hypothesis: str = Field(default="", max_length=2000)
    rollback_action_ref: str = Field(default="", max_length=240)
    proposed_fix: str = Field(default="", max_length=2000)
    status: IncidentStatus = "open"
    channel_scope: list[str] = Field(default_factory=list, max_length=20)
    notification_targets: list[str] = Field(default_factory=list, max_length=10)
    created_from: str = Field(default="manual", max_length=120)


class IncidentTransition(BaseModel):
    model_config = ConfigDict(extra="forbid")

    note: str = Field(default="", max_length=1200)


class IncidentRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    workspace_id: UUID
    agent_id: UUID
    deployment_id: str
    severity: IncidentSeverity
    trigger: str
    status: IncidentStatus
    affected_trace_ids: list[str]
    affected_conversation_count: int
    root_cause_hypothesis: str
    rollback_action_ref: str
    proposed_fix: str
    candidate_eval_suite_id: str | None
    fix_change_package_id: str | None = None
    channel_scope: list[str]
    notifications: list[dict[str, Any]]
    timeline: list[dict[str, Any]]
    report: dict[str, Any]
    created_at: datetime
    created_by: str
    resolved_at: datetime | None = None


def _timeline_event(kind: str, at: datetime, summary: str) -> dict[str, str]:
    return {"kind": kind, "at": at.isoformat(), "summary": summary}


def _notifications_for(body: IncidentCreate, now: datetime) -> list[dict[str, str]]:
    recipients = list(
        dict.fromkeys(target.strip() for target in body.notification_targets if target.strip())
    )
    return [
        {
            "recipient": recipient,
            "channel": "in_app",
            "status": "queued",
            "sent_at": now.isoformat(),
            "summary": f"{body.severity} incident: {body.trigger}",
        }
        for recipient in recipients
    ]


def _report_for(body: IncidentCreate, notifications: list[dict[str, Any]]) -> dict[str, Any]:
    return {
        "trigger": body.trigger,
        "affected_conversations": body.affected_conversation_count,
        "affected_channels": body.channel_scope,
        "customer_impact": (
            "Unknown until affected traces are reviewed."
            if body.affected_conversation_count == 0
            else f"{body.affected_conversation_count} conversations require review."
        ),
        "actions_taken": [body.rollback_action_ref] if body.rollback_action_ref else [],
        "suspected_cause": body.root_cause_hypothesis
        or "Pending trace and deployment-event review.",
        "proposed_fix": body.proposed_fix or "Generate candidate evals before fix.",
        "candidate_regression_tests": body.affected_trace_ids,
        "rollback_status": "executed" if body.rollback_action_ref else "not_started",
        "notifications": notifications,
    }


def incident_payload(record: IncidentRecord) -> dict[str, Any]:
    return record.model_dump(mode="json")


class IncidentRegistry:
    def __init__(self) -> None:
        self._incidents: dict[UUID, list[IncidentRecord]] = {}
        self._lock = asyncio.Lock()

    async def list_for_workspace(
        self,
        *,
        workspace_id: UUID,
        agent_id: UUID | None = None,
    ) -> list[IncidentRecord]:
        async with self._lock:
            records = list(self._incidents.get(workspace_id, []))
        if agent_id is not None:
            records = [record for record in records if record.agent_id == agent_id]
        return records

    async def create(
        self,
        *,
        agent: AgentRecord,
        body: IncidentCreate,
        actor_sub: str,
    ) -> IncidentRecord:
        async with self._lock:
            now = datetime.now(UTC)
            notifications = _notifications_for(body, now)
            record = IncidentRecord(
                id=f"inc_{uuid4().hex[:12]}",
                workspace_id=agent.workspace_id,
                agent_id=agent.id,
                deployment_id=body.deployment_id,
                severity=body.severity,
                trigger=body.trigger,
                status=body.status,
                affected_trace_ids=body.affected_trace_ids,
                affected_conversation_count=body.affected_conversation_count,
                root_cause_hypothesis=body.root_cause_hypothesis,
                rollback_action_ref=body.rollback_action_ref,
                proposed_fix=body.proposed_fix,
                candidate_eval_suite_id=None,
                fix_change_package_id=None,
                channel_scope=body.channel_scope,
                notifications=notifications,
                timeline=[
                    _timeline_event(
                        "incident_created",
                        now,
                        f"Incident created from {body.created_from}.",
                    )
                ],
                report=_report_for(body, notifications),
                created_at=now,
                created_by=actor_sub,
            )
            self._incidents.setdefault(agent.workspace_id, []).insert(0, record)
            return record

    async def create_for_rollback(
        self,
        *,
        agent: AgentRecord,
        deployment_id: str,
        version_id: str,
        actor_sub: str,
        mode: Literal["manual", "auto"] = "manual",
        trigger: str = "",
        reason: str = "",
        affected_trace_ids: list[str] | None = None,
        notification_targets: list[str] | None = None,
    ) -> IncidentRecord:
        label = "auto-rollback" if mode == "auto" else "manual rollback"
        trace_ids = affected_trace_ids or []
        body = IncidentCreate(
            deployment_id=deployment_id,
            severity="high" if mode == "auto" else "medium",
            trigger=trigger or f"{label} executed for deployment {deployment_id}",
            affected_trace_ids=trace_ids,
            affected_conversation_count=len(trace_ids),
            root_cause_hypothesis=reason or "Rollback executed before root cause was confirmed.",
            rollback_action_ref=f"deployment/{deployment_id}/rollback",
            proposed_fix=(
                f"Review traces around deployment {deployment_id}, seed incident "
                f"evals, then create a Change Package against {version_id}."
            ),
            status="contained",
            notification_targets=notification_targets or [],
            created_from=label.replace(" ", "_"),
        )
        record = await self.create(agent=agent, body=body, actor_sub=actor_sub)
        now = datetime.now(UTC)
        contained = record.model_copy(
            update={
                "timeline": [
                    *record.timeline,
                    *(
                        [
                            _timeline_event(
                                "affected_traces_collected",
                                now,
                                f"{len(trace_ids)} affected trace(s) attached to the incident report.",
                            )
                        ]
                        if trace_ids
                        else []
                    ),
                    _timeline_event(
                        "containment",
                        now,
                        f"{label.capitalize()} moved traffic away from {version_id}.",
                    ),
                ]
            }
        )
        await self._replace(contained)
        return contained

    async def get(self, *, agent: AgentRecord, incident_id: str) -> IncidentRecord:
        async with self._lock:
            for record in self._incidents.get(agent.workspace_id, []):
                if record.agent_id == agent.id and record.id == incident_id:
                    return record
        raise WorkspaceError(f"unknown incident: {incident_id}")

    async def transition(
        self,
        *,
        agent: AgentRecord,
        incident_id: str,
        status: IncidentStatus,
        note: str,
    ) -> IncidentRecord:
        record = await self.get(agent=agent, incident_id=incident_id)
        now = datetime.now(UTC)
        updated = record.model_copy(
            update={
                "status": status,
                "resolved_at": now if status == "resolved" else record.resolved_at,
                "timeline": [
                    *record.timeline,
                    _timeline_event(
                        status,
                        now,
                        note or f"Incident moved to {status}.",
                    ),
                ],
            }
        )
        await self._replace(updated)
        return updated

    async def link_eval_suite(
        self,
        *,
        agent: AgentRecord,
        incident_id: str,
        suite_id: str,
    ) -> IncidentRecord:
        record = await self.get(agent=agent, incident_id=incident_id)
        now = datetime.now(UTC)
        updated = record.model_copy(
            update={
                "candidate_eval_suite_id": suite_id,
                "timeline": [
                    *record.timeline,
                    _timeline_event(
                        "evals_seeded",
                        now,
                        f"Candidate regression evals added to suite {suite_id}.",
                    ),
                ],
                "report": {
                    **record.report,
                    "candidate_eval_suite_id": suite_id,
                },
            }
        )
        await self._replace(updated)
        return updated

    async def link_fix_change_package(
        self,
        *,
        agent: AgentRecord,
        incident_id: str,
        package_id: str,
    ) -> IncidentRecord:
        record = await self.get(agent=agent, incident_id=incident_id)
        now = datetime.now(UTC)
        updated = record.model_copy(
            update={
                "status": "fix_staged",
                "fix_change_package_id": package_id,
                "timeline": [
                    *record.timeline,
                    _timeline_event(
                        "fix_change_package_created",
                        now,
                        f"Fix Change Package {package_id} created from incident.",
                    ),
                ],
                "report": {
                    **record.report,
                    "fix_change_package_id": package_id,
                    "actions_taken": [
                        *record.report.get("actions_taken", []),
                        f"change_package/{package_id}",
                    ],
                },
            }
        )
        await self._replace(updated)
        return updated

    async def _replace(self, updated: IncidentRecord) -> None:
        async with self._lock:
            records = self._incidents.get(updated.workspace_id, [])
            for index, record in enumerate(records):
                if record.id == updated.id:
                    records[index] = updated
                    return
            raise WorkspaceError(f"unknown incident: {updated.id}")

File: apps/studio/src/lib/incidents.ts
import { cpJson, type UxWireupClientOptions } from "@/lib/ux-wireup";
import type { ChangePackage } from "@/lib/change-package";

export type IncidentSeverity = "low" | "medium" | "high" | "critical";
export type IncidentStatus =
  | "open"
  | "contained"
  | "investigating"
  | "fix_staged"
  | "resolved"
  | "archived";

export interface IncidentRecord {
  id: string;
  workspace_id: string;
  agent_id: string;
  deployment_id: string;
  severity: IncidentSeverity;
  trigger: string;
  status: IncidentStatus;
  affected_trace_ids: string[];
  affected_conversation_count: number;
  root_cause_hypothesis: string;
  rollback_action_ref: string;
  proposed_fix: string;
  candidate_eval_suite_id: string | null;
  fix_change_package_id: string | null;
  channel_scope: string[];
  notifications: Array<{
    recipient: string;
    channel: string;
    status: string;
    sent_at: string;
    summary: string;
  }>;
  timeline: Array<{ kind: string; at: string; summary: string }>;
  report: Record<string, unknown>;
  created_at: string;
  created_by: string;
  resolved_at: string | null;
}

export interface IncidentListResponse {
  items: IncidentRecord[];
}

export interface IncidentEvalSeedResponse {
  ok: boolean;
  suite_id: string;
  case_ids: string[];
  incident: IncidentRecord;
}

export interface IncidentFixPackageResponse {
  ok: boolean;
  change_package: ChangePackage;
  incident: IncidentRecord;
}

export type IncidentTransitionAction =
  | "contain"
  | "investigate"
  | "resolve"
  | "archive";

const LOCAL_INCIDENTS: IncidentRecord[] = [
  {
    id: "inc_local_rollback",
    workspace_id: "local",
    agent_id: "agent_support",
    deployment_id: "dep_local",
    severity: "high",
    trigger: "error_rate breached 4% for web_chat canary",
    status: "contained",
    affected_trace_ids: ["trace_refund_742"],
    affected_conversation_count: 17,
    root_cause_hypothesis: "Tool schema changed upstream during canary.",
    rollback_action_ref: "deployment/dep_local/rollback",
    proposed_fix:
      "Replay affected traces, pin the tool schema, and create a Change Package with regression coverage.",
    candidate_eval_suite_id: null,
    fix_change_package_id: null,
    channel_scope: ["web_chat"],
    notifications: [
      {
        recipient: "maya@acme.test",
        channel: "in_app",
        status: "queued",
        sent_at: "2026-05-09T03:12:00.000Z",
        summary: "high incident: error_rate breached 4% for web_chat canary",
      },
      {
        recipient: "diego@acme.test",
        channel: "in_app",
        status: "queued",
        sent_at: "2026-05-09T03:12:00.000Z",
        summary: "high incident: error_rate breached 4% for web_chat canary",
      },
    ],
    timeline: [
      {
        kind: "incident_created",
        at: "2026-05-09T03:12:00.000Z",
        summary: "Incident created from auto_rollback.",
      },
      {
        kind: "containment",
        at: "2026-05-09T03:12:38.000Z",
        summary: "Auto-rollback moved traffic away from v24.",
      },
    ],
    report: {
      rollback_status: "executed",
      suspected_cause: "Tool schema changed upstream during canary.",
      proposed_fix:
        "Replay affected traces, pin the tool schema, and create a Change Package with regression coverage.",
    },
    created_at: "2026-05-09T03:12:00.000Z",
    created_by: "loop-system",
    resolved_at: null,
  },
];

function getLocalIncident(agentId: string, incidentId: string): IncidentRecord {
  const incident =
    LOCAL_INCIDENTS.find((item) => item.id === incidentId) ??
    LOCAL_INCIDENTS[0]!;
  return {
    ...incident,
    id: incidentId,
    agent_id: agentId,
  };
}

export async function listWorkspaceIncidents(
  workspaceId: string,
  opts: UxWireupClientOptions = {},
): Promise<IncidentListResponse> {
  return cpJson<IncidentListResponse>(
    `/workspaces/${encodeURIComponent(workspaceId)}/incidents`,
    {
      ...opts,
      fallback: { items: LOCAL_INCIDENTS },
    },
  );
}

export async function listAgentIncidents(
  agentId: string,
  opts: UxWireupClientOptions = {},
): Promise<IncidentListResponse> {
  return cpJson<IncidentListResponse>(
    `/agents/${encodeURIComponent(agentId)}/incidents`,
    {
      ...opts,
      fallback: {
        items: LOCAL_INCIDENTS.filter(
          (incident) => incident.agent_id === agentId,
        ),
      },
    },
  );
}

export async function seedIncidentEvalCases(
  agentId: string,
  incidentId: string,
  opts: UxWireupClientOptions = {},
): Promise<IncidentEvalSeedResponse> {
  const localIncident = getLocalIncident(agentId, incidentId);
  return cpJson<IncidentEvalSeedResponse>(
    `/agents/${encodeURIComponent(agentId)}/incidents/${encodeURIComponent(
      incidentId,
    )}/eval-cases`,
    {
      ...opts,
      method: "POST",
      fallback: {
        ok: true,
        suite_id: "suite_incident_regressions_local",
        case_ids: [`case_${incidentId}`],
        incident: {
          ...localIncident,
          candidate_eval_suite_id: "suite_incident_regressions_local",
        },
      },
    },
  );
}

export async function createIncidentFixChangePackage(
  agentId: string,
  incidentId: string,
  opts: UxWireupClientOptions = {},
): Promise<IncidentFixPackageResponse> {
  const localIncident = getLocalIncident(agentId, incidentId);
  const now = new Date().toISOString();
  return cpJson<IncidentFixPackageResponse>(
    `/agents/${encodeURIComponent(agentId)}/incidents/${encodeURIComponent(
      incidentId,
    )}/change-package`,
    {
      ...opts,
      method: "POST",
      body: {},
      fallback: {
        ok: true,
        change_package: {
          id: `cp_${incidentId}`,
          workspace_id: localIncident.workspace_id,
          agent_id: agentId,
          branch_id: "incident/fix",
          change_set_id: `cs_${incidentId}`,
          release_candidate_id: `rc_${incidentId}`,
          from_version_id: "production",
          to_version_id: "draft-incident-fix",
          commitment_document_id: "commitment_local",
          commitment_document_version: 1,
          summary: `Fix incident ${localIncident.id}: ${localIncident.trigger}`,
          semantic_diff: [
            {
              dimension: "incident",
              summary: localIncident.proposed_fix,
              evidence_ref: `incident/${localIncident.id}`,
            },
          ],
          eval_results_ref:
            localIncident.candidate_eval_suite_id ??
            `incident/${localIncident.id}/candidate-evals`,
          replay_results_ref: `incident/${localIncident.id}/affected-traces`,
          risk_summary: `${localIncident.severity} incident fix.`,
          cost_summary: "Replay required before cost claims.",
          latency_summary: "Replay required before latency claims.",
          channel_readiness_summary: `Incident channel scope: ${localIncident.channel_scope.join(
            ", ",
          )}`,
          tool_changes: [],
          memory_changes: [],
          knowledge_changes: [],
          required_approvals: [],
          pre_approved_classes: [],
          approval_status: "blocked",
          rollback_target_version_id:
            localIncident.rollback_action_ref || "last-known-safe",
          evidence_pack_id: `ep_${incidentId}`,
          evidence: {
            incident: localIncident.id,
            replay_results: `incident/${localIncident.id}/affected-traces`,
          },
          content_hash: `local-${incidentId}`,
          status: "generated",
          created_at: now,
          updated_at: now,
          submitted_at: null,
          stale_at: null,
        },
        incident: {
          ...localIncident,
          status: "fix_staged",
          fix_change_package_id: `cp_${incidentId}`,
          report: {
            ...localIncident.report,
            fix_change_package_id: `cp_${incidentId}`,
          },
        },
      },
    },
  );
}

export async function transitionIncident(
  agentId: string,
  incidentId: string,
  action: IncidentTransitionAction,
  note: string,
  opts: UxWireupClientOptions = {},
): Promise<IncidentRecord> {
  const localIncident = getLocalIncident(agentId, incidentId);
  const statusByAction: Record<IncidentTransitionAction, IncidentStatus> = {
    contain: "contained",
    investigate: "investigating",
    resolve: "resolved",
    archive: "archived",
  };
  const nextStatus = statusByAction[action];
  const now = new Date().toISOString();
  return cpJson<IncidentRecord>(
    `/agents/${encodeURIComponent(agentId)}/incidents/${encodeURIComponent(
      incidentId,
    )}/${action}`,
    {
      ...opts,
      method: "POST",
      body: { note },
      fallback: {
        ...localIncident,
        status: nextStatus,
        resolved_at:
          nextStatus === "resolved" ? now : localIncident.resolved_at,
        timeline: [
          ...localIncident.timeline,
          {
            at: now,
            kind: `incident_${nextStatus}`,
            summary: note || `Incident moved to ${nextStatus}.`,
          },
        ],
      },
    },
  );
}
